### PR TITLE
Blueprints Rework

### PR DESCRIPTION
**Pull request by RustingWithYou**
Reworks blueprints to use an eye component for area selection and changing. Blueprints will now work on whichever overmap site they spawn on if overmap is enabled, though currently they have only been added to the Horizon.

Adds shuttle-modifying blueprints for altering shuttle areas and exoplanet outpost blueprints for creating areas on exoplanets. A set of outpost blueprints and a lockbox containing blueprints for the Horizon's shuttles have been added to the CE's locker.

Moves eye creation to a component.

Ported from:
https://github.com/NebulaSS13/Nebula/pull/465
https://github.com/NebulaSS13/Nebula/pull/564
https://github.com/NebulaSS13/Nebula/pull/3046

### CODE REVIEW

**Comment by RustingWithYou** [comment]
!review

**Comment by RustingWithYou** [comment]
!review

File: code/datums/observation/logged_out.dm [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+//		Raised when: A mob logs out (client either logged out or was moved to another mob)
+//
+//		Arguments that the called proc should expect:
+//			/mob/leaver:    The mob that has logged out
+//			/client/client: The mob's client
+
+GLOBAL_DATUM_INIT(logged_out_event, /singleton/observ/logged_out, new)
+
+/singleton/observ/logged_out
+	name = "Logged Out"
```
Comment (NonQueueingMatt) [inline code/datums/observation/logged_out.dm:13]: signals not singletons

File: code/datums/observation/logged_in.dm [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+//		Raised when: A mob logs in (not client)
+//
+//		Arguments that the called proc should expect:
+//			/mob/joiner: The mob that has logged in
+
+GLOBAL_DATUM_INIT(logged_in_event, /singleton/observ/logged_in, new)
+
+/singleton/observ/logged_in
+	name = "Logged In"
+	expected_type = /mob
```
Comment (NonQueueingMatt) [inline code/datums/observation/logged_in.dm:13]: signals not singletons

**Review by NonQueueingMatt — CHANGES_REQUESTED** [review]

File: code/_onclick/hud/action.dm [hunk 3]
```diff
@@ -10,6 +10,9 @@
 #define AB_CHECK_ALIVE 8
 #define AB_CHECK_INSIDE 16
 
+#define PARENT_TARGET		0
+#define EYE_TARGET			1
+#define COMPONENT_TARGET	2
```
Comment (FluffyGhoster) [inline code/_onclick/hud/action.dm:15]: DMDoc

File: code/_onclick/hud/action.dm [hunk 4]
```diff
@@ -288,6 +291,27 @@
 	to_chat(usr, SPAN_NOTICE("You press the button on the exterior of \the [target_clothing]."))
 	target_clothing.action_circuit.activate_pin(1)
 
+/datum/action/eye
+	action_type = AB_GENERIC
+	check_flags = AB_CHECK_LYING|AB_CHECK_STUNNED
+	var/eye_type = /mob/abstract/eye
+	var/target_type = PARENT_TARGET //The relevant owner of the proc to be called by the action.
```
Comment (FluffyGhoster) [inline code/_onclick/hud/action.dm:298]: DMDoc

File: code/_onclick/hud/action.dm [hunk 5]
```diff
@@ -288,6 +291,27 @@
...
+
+/datum/action/eye/New(var/datum/component/eye/eye_component)
+	switch(target_type)
```
Comment (FluffyGhoster) [inline code/_onclick/hud/action.dm:301]: Verify that the eye_component is of appropriate type on creation before going forward, with !istype() -> crashwith

File: code/datums/components/eye/_eye.dm [hunk 6]
```diff
@@ -0,0 +1,84 @@
+/datum/component/eye
+	///The actual eye mob linked to the component
+	var/mob/abstract/eye/component_eye
+	///What type of eye is used
+	var/eye_type = /mob/abstract/eye
+	///The mob currently looking through the eye
+	var/mob/current_looker
+
+	//Actions used to pass commands from the eye to the holder. Must be subtypes of /datum/action/eye
```
Comment (FluffyGhoster) [inline code/datums/components/eye/_eye.dm:9]:
```suggestion
	///Actions used to pass commands from the eye to the holder. Must be subtypes of /datum/action/eye
```

File: code/datums/components/eye/_eye.dm [hunk 7]
```diff
@@ -0,0 +1,84 @@
+/datum/component/eye
+	///The actual eye mob linked to the component
+	var/mob/abstract/eye/component_eye
```
Comment (FluffyGhoster) [inline code/datums/components/eye/_eye.dm:3]: Null or QDEL_NULL this reference on Destroy

File: code/datums/components/eye/_eye.dm [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+	///What type of eye is used
+	var/eye_type = /mob/abstract/eye
+	///The mob currently looking through the eye
+	var/mob/current_looker
```
Comment (FluffyGhoster) [inline code/datums/components/eye/_eye.dm:7]: Null this reference on Destroy

File: code/datums/observation/logged_in.dm [hunk 9]
```diff
@@ -0,0 +1,21 @@
+//	Observer Pattern Implementation: Logged in
+//		Registration type: /mob
+//
+//		Raised when: A mob logs in (not client)
+//
+//		Arguments that the called proc should expect:
+//			/mob/joiner: The mob that has logged in
+
+GLOBAL_DATUM_INIT(logged_in_event, /singleton/observ/logged_in, new)
```
Comment (FluffyGhoster) [inline code/datums/observation/logged_in.dm:9]: No observables, use signals

File: code/datums/observation/logged_out.dm [hunk 10]
```diff
@@ -0,0 +1,22 @@
+//	Observer Pattern Implementation: Logged Out
+//		Registration type: /mob
+//
+//		Raised when: A mob logs out (client either logged out or was moved to another mob)
+//
+//		Arguments that the called proc should expect:
+//			/mob/leaver:    The mob that has logged out
+//			/client/client: The mob's client
+
+GLOBAL_DATUM_INIT(logged_out_event, /singleton/observ/logged_out, new)
```
Comment (FluffyGhoster) [inline code/datums/observation/logged_out.dm:10]: No observables, use signals

File: code/modules/mob/abstract/freelook/blueprints/blueprints.dm [hunk 11]
```diff
@@ -0,0 +1,326 @@
...
+	var/list/selection_images = list()
+	var/turf/last_selected_turf
+	var/image/last_selected_image
+
+	///On what z-levels the blueprints can be used to modify or create areas.
+	var/list/valid_z_levels = list()
+
+	///Displayed to the user to allow them to see what area they're hovering over
+	var/obj/effect/overlay/area_name_effect
+	var/area_prefix
```
Comment (FluffyGhoster) [inline code/modules/mob/abstract/freelook/blueprints/blueprints.dm:15]: DMDoc

File: code/modules/mob/abstract/freelook/blueprints/blueprints.dm [hunk 12]
```diff
@@ -0,0 +1,326 @@
+#define MAX_AREA_SIZE 300
+
+/mob/abstract/eye/blueprints
+	/// Associative list of turfs -> boolean validity that the player has selected for new area creation.
+	var/list/selected_turfs = list()
+	var/list/selection_images = list()
+	var/turf/last_selected_turf
+	var/image/last_selected_image
```
Comment (FluffyGhoster) [inline code/modules/mob/abstract/freelook/blueprints/blueprints.dm:8]: DMDoc

File: code/modules/mob/abstract/freelook/blueprints/blueprints.dm [hunk 13]
```diff
@@ -0,0 +1,326 @@
...
+	area_name_effect.plane = HUD_PLANE
+	area_name_effect.appearance_flags = APPEARANCE_UI_IGNORE_ALPHA
+	area_name_effect.screen_loc = "LEFT+1,BOTTOM+2"
+
+	last_selected_image = image('icons/effects/blueprints.dmi', "selected")
+	last_selected_image.plane = HUD_PLANE
+	last_selected_image.appearance_flags = NO_CLIENT_COLOR|RESET_COLOR
+
+/mob/abstract/eye/blueprints/Destroy()
+	. = ..()
```
Comment (FluffyGhoster) [inline code/modules/mob/abstract/freelook/blueprints/blueprints.dm:40]: Call parent last in destroy, unless you have a good reason not to

File: code/datums/observation/logged_out.dm (hunk 10 above)
Comment (RustingWithYou) [reply code/datums/observation/logged_out.dm:10]: Forgot that I had this in from the click handler PR, will remove shortly

File: code/datums/components/eye/_eye.dm (hunk 7 above)
Comment (RustingWithYou) [reply code/datums/components/eye/_eye.dm:3]: It's QDEL_NULLed on unlook which is called on Destroy

File: code/datums/components/eye/_eye.dm (hunk 8 above)
Comment (RustingWithYou) [reply code/datums/components/eye/_eye.dm:7]: Same as above

**Review by NonQueueingMatt — APPROVED** [review]

**Review by FluffyGhoster — APPROVED** [review]
It won't destroy the server.
Probably.

**Pull request merged**